### ISSUE TITLE
Joint test session: OASYS vs LibWiser

### ISSUE THREAD

**Issue by capitanevs** [oasys-elettra-kit/OASYS1-Wiser#52]
Here I test the following beamline.

![immagine](https://user-images.githubusercontent.com/13431660/123123620-4298aa80-d447-11eb-828e-bde59d9c5354.png)

The first results  I get are:

LibWiser:
![immagine](https://user-images.githubusercontent.com/13431660/123123688-517f5d00-d447-11eb-9229-3a4a5885b64e.png)

Oasys:
![immagine](https://user-images.githubusercontent.com/13431660/123123730-59d79800-d447-11eb-9478-cbee83838b2e.png)

*So, there is a problem in positioning*.

Let's have a look at the widget settings.... (next)

**Comment by capitanevs** [comment]
I found one mistake I did in  the focussing mirror:
![immagine](https://user-images.githubusercontent.com/13431660/123124091-aa4ef580-d447-11eb-8502-c86d64ee397e.png)

If I manually set the distance, I get an "almost right result"
LibWiser:
![immagine](https://user-images.githubusercontent.com/13431660/123124453-fb5ee980-d447-11eb-8475-adbe3789da02.png)


Oasys:
![immagine](https://user-images.githubusercontent.com/13431660/123124255-ceaad200-d447-11eb-99d1-cc8c22abf09e.png)

QUITE good, but... LibWiser has a wrong scale?

**Comment by capitanevs** [comment]
`fm_h.PlotIntensity()` return the right plot

![immagine](https://user-images.githubusercontent.com/13431660/123127446-95279600-d44a-11eb-9416-922e3cf23e30.png)


`Beamline.PlotIntensity()`
does not 

![immagine](https://user-images.githubusercontent.com/13431660/123127503-a07ac180-d44a-11eb-96e7-c4dc45d87c0e.png)


I'll fix this...

**Comment by capitanevs** [comment]
Plot bug corrected . 
![immagine](https://user-images.githubusercontent.com/13431660/123128111-27c83500-d44b-11eb-9c00-aab1bae3dd44.png)
![immagine](https://user-images.githubusercontent.com/13431660/123128194-3a426e80-d44b-11eb-9bb3-aa4e624834b9.png)

![immagine](https://user-images.githubusercontent.com/13431660/123128207-3d3d5f00-d44b-11eb-9e15-5e421094f6b1.png)

**Comment by capitanevs** [comment]
Also the focusing on the detector now is correct:

![immagine](https://user-images.githubusercontent.com/13431660/123128349-5e05b480-d44b-11eb-9b7f-104b83c2f6af.png)

**Comment by capitanevs** [comment]
I close, update and save.

I get errors at the detector:
![immagine](https://user-images.githubusercontent.com/13431660/123134625-492c1f80-d451-11eb-8d2f-641ae5824b93.png)

![immagine](https://user-images.githubusercontent.com/13431660/123134653-4f220080-d451-11eb-9370-ca1c4da2cf40.png)

On problem is the mu character of enf formatting (omg).
I fix this
![immagine](https://user-images.githubusercontent.com/13431660/123135004-a32ce500-d451-11eb-83b1-8978e02a36bf.png)


However there must be another problem somewhere else. Look at the intensity on the KB.

I see that the distances of slits and KB are WRONG:

![immagine](https://user-images.githubusercontent.com/13431660/123135278-e6875380-d451-11eb-9a5b-7c9d7ccd65a2.png)
![immagine](https://user-images.githubusercontent.com/13431660/123135319-f141e880-d451-11eb-9880-12159732b8e4.png)


**ARE WE SURE THAT UPDATING DOES NOT AFFECT SAVED PARAMETERS?** (old question)

**Comment by capitanevs** [comment]
**The widget of KB has been defaulted!**
![immagine](https://user-images.githubusercontent.com/13431660/123135776-775e2f00-d452-11eb-8ae3-3d4efa465121.png)

before F1 and F2 were 99 and 1.4 and distance was 99.

**investigate**

**Comment by capitanevs** [comment]
also the names have been changed
![immagine](https://user-images.githubusercontent.com/13431660/123135890-9d83cf00-d452-11eb-8e0d-7a8e5434d4f5.png)

**Comment by capitanevs** [comment]
I do "save",
close Oasys, 
open 
test_wiser_margot_correct3.ows

**Comment by capitanevs** [comment]
**Conclusions**

- Before going on, we should get rid of the "data not plottable". Also, sealing the plotting procedure with try and except could be a good idea. **Ideal behavior**: even if data are not plottable, there is an out_object.
- When everything is fine, Oasys and LibWiser return the same results (as expected)
- updating the code create issues with the stored data, sometimes. It is still not clear WHEN, but for sure IT HAPPENS
- We still have to debug the figure error. There is no problem with LibWiser.
- I have created a script for quickly inspecting the beamline at a glance.

**Comment by capitanevs** [comment]
[test_wiser_margot_correct3.zip](https://github.com/oasys-elettra-kit/OASYS1-Wiser/files/6703652/test_wiser_margot_correct3.zip)
just rename zip into ows

**Comment by capitanevs** [comment]
** Now, together wi Aljosa**

file: margot3,
before update
![immagine](https://user-images.githubusercontent.com/13431660/123227307-bfbe3100-d4d4-11eb-88f1-0a2aa6ed1c7d.png)

**Comment by capitanevs** [comment]
![immagine](https://user-images.githubusercontent.com/13431660/123231847-df575880-d4d8-11eb-950d-cbca693e41a9.png)

**Comment by capitanevs** [comment]
About Numba:
![immagine](https://user-images.githubusercontent.com/13431660/123236961-8938e400-d4dd-11eb-8d42-97399c174619.png)

**Comment by aljosahafner** [comment]
* Please push the fixed version of Units
* Does this help, after Numba update (seems to be the source of previous errors) I am getting this error with FigureError 
![image](https://user-images.githubusercontent.com/51963146/123240881-ff8b1580-d4e0-11eb-9d61-ef2acdcfaa6d.png)

**Comment by capitanevs** [comment]
**Intermediate  Conclusions**

- Figure error DID Work on Michele's laptop (without any apparent reason, but subsequent upgrades) (*)
- There IS an error in plotting "mu" unicode character. This is regards ALL the widgets (not the detector only, as guessed at a certain point), and it must be somehow related to the Units.py. We can say this because if the optical element il large enough, everything is fine, but if it comes to um range, then the error arises.
- Aljosa and his IDE/libraries have quarreled, and it will take time to reconcile :-)  
- SRW<-> Oasys Link are working.

**Comment by capitanevs** [comment]
> 
> 
>     * Please push the fixed version of Units
> 
>     * Does this help, after Numba update (seems to be the source of previous errors) I am getting this error with FigureError
>       ![image](https://user-images.githubusercontent.com/51963146/123240881-ff8b1580-d4e0-11eb-9d61-ef2acdcfaa6d.png)

Interesting and curious.

I obviously conceived the message because it often happened that one of the two arrays was 0 length (itis what it is written).
But even more obviously this is not what is happening: they are 5000 and 5016 long. np.linspace seems to be a good candidate

I will check.

Can you share the very same file?

**Comment by aljosahafner** [comment]
Definitely the source of the "data not plottable" is Units package! If I remove the usage of GetAxisSI out, then the error dissapears!

**Comment by capitanevs** [comment]
> 
> 
> Definitely the source of the "data not plottable" is Units package! If I remove the usage of GetAxisSI out, then the error dissapears!

Yes, but how? GetAxisSI is returning two values: an array (which does no contain strings) and a character, which we hardcoded to "K". There must be another link to Units.py.....

**Comment by aljosahafner** [comment]
I sent you an E-mail, maybe this helps with debugging...
![image](https://user-images.githubusercontent.com/51963146/123243616-79bc9980-d4e3-11eb-845d-3c2dbc79ffb6.png)

**Comment by capitanevs** [comment]
I am confused... this is what I get now with the same lines...
![immagine](https://user-images.githubusercontent.com/13431660/123244256-06675780-d4e4-11eb-9745-8a9bf9f31b65.png)

**Comment by aljosahafner** [comment]
![image](https://user-images.githubusercontent.com/51963146/123245224-e8e6bd80-d4e4-11eb-9cc3-0de7102b88d0.png)

**Comment by capitanevs** [comment]
When you have time, let's try to run this test code

```
#%%
from LibWiser import Units
import numpy as np
Units.PrefixLookup['u']
Units.PrefixLookup['μ']

x = 2e-6
xx = np.array([x, 2*x])
Units.GetEngFormatWithPrefix(x)
#Units.GetEngFormatWithExponential(x)
Units.GetEngArgument(x)
Units.GetEngPrefix(x)
Units.GetEngInfo(xx)
Units.GetEngAxis(xx)
Units.GetAxisSI(xx)

```

there's my output
![immagine](https://user-images.githubusercontent.com/13431660/123246761-6eb73880-d4e6-11eb-9845-e287a16fb32d.png)

**Comment by aljosahafner** [comment]
![image](https://user-images.githubusercontent.com/51963146/123247070-c655a400-d4e6-11eb-9a75-75d059929d36.png)

**Comment by aljosahafner** [comment]
![image](https://user-images.githubusercontent.com/51963146/123247466-39f7b100-d4e7-11eb-882e-7402807f4d23.png)

**Comment by capitanevs** [comment]
Ok, that IS really funny...


![immagine](https://user-images.githubusercontent.com/13431660/123247800-893de180-d4e7-11eb-8940-ac8d8a064bf1.png)

Going to lunch now, back on it late....

**Comment by capitanevs** [comment]
I have pushed my latest version of units

**Comment by aljosahafner** [comment]
Yes, it's the character, somehow you can't pass it around, probably some encoding problem... Let's try to avoid this character entirely

**Comment by capitanevs** [comment]
This is even more funny

![immagine](https://user-images.githubusercontent.com/13431660/123265546-759c7600-d4fb-11eb-8d21-82def453e072.png)

**Comment by capitanevs** [comment]
Ok, the mu is now totally disappeared.

**Comment by aljosahafner** [comment]
Excellent! Plotting works, starting from `v0.2.29`!

**Comment by aljosahafner** [comment]
What remains to be solved from this test session is the FigureError.
